src/main/java/build/buildfarm/common/services/WriteStreamObserver.java

```diff
@@ -403,6 +405,11 @@ private void writeData(ByteString data) throws EntryLimitException {
       ioMetric.observe(data.size());
     } catch (EntryLimitException e) {
       throw e;
+    } catch (WriteCompleteException e) {
+      Status status = Status.fromCode(ALREADY_EXISTS)
+          .withCause(e);
+      log.log(Level.FINE, format("already wrote data for %s", name), e);
+      errorResponse(status.asException());
```

Comment:
Conditional this fine logging with `if (errorResponse...` to be complete - errorResponse returns false if we've already delivered a response to the stream.

Reply:
SGTM - I've wrapped it in this conditional 👍